src/string_scanner.cr

```diff
@@ -125,6 +148,51 @@ class StringScanner
     end
   end
 
+  private def match(pattern : String, advance = true, anchored = true)
+    @last_match = nil
+    if pattern.bytesize > @str.bytesize - @byte_offset
+      nil
+    elsif anchored
+      if @str.unsafe_byte_slice(@byte_offset, pattern.bytesize) == pattern.unsafe_byte_slice(0)
+        @last_match = StringMatchData.new(pattern)
+        @byte_offset += pattern.bytesize if advance
+        pattern
+      end
+    elsif (found = @str.byte_index(pattern, @byte_offset))
+      finish = found + pattern.bytesize
+      result = @str.byte_slice(@byte_offset, finish - @byte_offset)
```

Comment:
If the pattern is a `String`, and `#byte_index` compares the arguments like opaque byte sequences (i.e. invalid UTF-8 byte sequences are not equivalent to U+FFFD), then `result` should always be equal to `pattern` and this line shouldn't be necessary. Ditto for the `Char` overload below

Reply:
It is for “scan_until”, so it won’t match pattern.